devito/dse/manipulation.py

```diff
@@ -284,14 +284,25 @@ def _topological_sort(exprs):
                 # Avoid cyclic dependences, such as
                 # Eq(f, f + 1)
                 continue
-            elif r.is_Indexed:
-                # Only scalars enforce an ordering
+            elif r.is_Indexed and e.lhs in retrieve_terminals(mapper[r].rhs):
+                # Cyclic dependencies stay "as is" ie
+                # [Eq(r1, u + 1), Eq(u, r1 + 2)]
                 continue
             else:
                 dag.add_edge(mapper[r], e, force_add=True)
     processed = dag.topological_sort()
 
     # Append tensor equations at the end in user-provided order
-    processed.extend(tensors)
+    processed_ordered = []
+    for p in processed:
+        # Tensor equation was required for the sorting, add all up to that
```

Comment:
This comment is unclear to me and could do with some re-wording.

Reply:
Tried to be clearer